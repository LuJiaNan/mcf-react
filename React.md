title: React开发总结和解决方案
author:
  name: jaxchow
output: react.html
controls: true

--

# React开发总结和改进

--

## 我们现有问题

  - 组件体系？各组件特性？
  - redux? 数据驱动 ?
  - 代码职责？组件化？模块化？
  - 规范 ？团队协作？
  - 自测？质量？

--

## 单向数据流

<img src="http://upload-images.jianshu.io/upload_images/16777-e9e03f0a1f0bc760.png?imageMogr2/auto-orient/" width="100%" >


--

## 前端组件化开发

> 最初的目的是代码重用，功能相对单一或者独立。在整个系统的代码层次上位于最底层，被其他代码所依赖，所以说组件化是纵向分层  

--

## 前端模块化开发

>最初的目的是将同一类型的代码整合在一起，所以模块的功能相对复杂，但都同属于一个业务。不同模块之间也会存在依赖关系，但大部分都是业务性的互相跳转，从地位上来说它们都是平级的

  * 简单化
  * 标准化
  * 可读性
  * 可维护性
  * 可扩展性
  * 可测试性  

--

## 两者区别


| 类别 |      目的  |   特点        | 接口   |       成果     | 架构定位 |
|-----|:---------:|--------------:|------:|----------------:|-------:|
| 组件 |  重用、解耦 | 高重用、松耦合 | 无统一接口 | 基础库、基础组件 |纵向分层 |
| 模块 |  隔离/封装 | 高内聚、松耦合 | 统一接口 | 业务框架、业务模块 | 横向分块 |

--

## 组件库 mcf-components

> componets组件库：提供自定义组件、ANTD 二次封装组件。


  - SearchFrom: 搜索组件
  - DataTable:  表格组件带分页和列配置
  - BaseForm :  基本Form 组件
  - ButtonGroups : 按钮组件库
  - FormItem :  表单项组件简化书写方式扩展功能
  - Permission : 权限展示组件，是否渲染
  - PropertyTable: 属性表格组件
  - TreeView ：树组件

--

## 业务组件库 mcf-crud

> crud 组件：为了解决日常业务开发｀增删改查｀功能抽象开发二次封装组件，减少大量重复性代码。


 - ListPage : 列表页的父类组件，提供日常的搜索与表格联动、行选参数据保存、公有的请求入口、页面跳转方法
 - FormPage : 表单页的父类组件，提供公有的请求入口、页面跳转方法、提交数据收集等。

--

## 装释器组件库 mcf-decorators

> 装释器实现多种组件的组合关系：能过高阶组件方式实现，必免代码之间的依赖问题


 - InnerComponent: 内并组件
 - NestedComponent: 嵌套组件
 - WrapperComponent: 外包组件

--

## 图标库 mcf-icons


>  基于ANTD 字体库的封装，可以与ANTD组件接合直接使用。自定义图标以 "mc-"为前缀定义。目前已整合了 审计、融合、工作流、数据库加密等库。

--

## 开发规范思路

  - 需求文档与业务接口先行
  - 代码分层化开发
  - 代码可测试
  - 模块间不依赖，所有共用采用独立包依赖
  - 组件可复用，共享组件使用
  - 团队或项目组一致性，差异化处理

--

## 相关资源


  `yarn add git+http://jaxlab.asuscomm.com/mcf/crud.git`
  `yarn add git+http://jaxlab.asuscomm.com/mcf/decorators.git`
  `yarn add git+http://jaxlab.asuscomm.com/mcf/icons.git`
  `yarn add git+http://jaxlab.asuscomm.com/mcf/components.git`
  [demo](http://jaxlab.asuscomm.com/mcf/router.git)
  [redux](https://redux.js.org/)
  [saga](https://redux-saga.js.org/)
