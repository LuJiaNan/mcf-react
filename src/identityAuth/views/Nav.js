import React from 'react'

export default class Nav extends React.Component{
	render(){
        return <a>首页</a>
    }
}
