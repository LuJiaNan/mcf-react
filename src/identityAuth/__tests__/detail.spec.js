import Detail from "../views/Detail.view";
import React from "react";
import { shallow, mount, render } from "enzyme";
import locale from "../locales";
import globalLocale from "../../../locales/zh-CN";
import Nav from "../views/Nav";
import PropTypes from "prop-types";
import renderer from "react-test-renderer"

function setup() {
  const props = {
    actions: {
      fetchItem: jest.fn(),
      saveListType: jest.fn(),
      changeAuthType: jest.fn(),
      changeListType: jest.fn(),
      handleFilter: jest.fn(),
      getFetchItem: jest.fn(),
      fetchDelete: jest.fn(),
      handlerMenu: jest.fn()
    },
    match: {
      params: {}
    },
    items:[],
    reducer: {
      type: ""
    },
    querys: jest.fn(),
    spins: jest.fn(),
    locale: function(key) {
      return locale[key] ? locale[key].defaultMessage : globalLocale[key];
    }
  };

  // const enzymeWrapper = shallow(<Detail {...props}/>,{disableLifecycleMethods: true})
  const shallowWrapper = shallow(<Detail {...props} a={1} />);
  const mountWrapper = mount(<Detail {...props} a={1} />);
  const renderWrapper = render(<Detail {...props} a={1} />);

  return {
    props,
    shallowWrapper,
    mountWrapper,
    renderWrapper
  };
}

describe("简单测试", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it.skip("综合测试", done => {
    let { shallowWrapper, props } = setup();

    // console.log(shallowWrapper.debug());
    // console.log(mountWrapper.debug());
    // console.dir(renderWrapper.find('button[actionKey="delete"]').prop("type"));
    // console.log(shallowWrapper.instance().props)
    // expect(shallowWrapper.instance().props.a).toEqual(1);

    // shallowWrapper.instance().changeAuthType()
    // expect(props.actions.changeListType.mock.calls.length).toEqual(1)
    // expect(props.actions.fetchItem.mock.calls.length).toBe(3)

    done();
  });

  it.skip("测试 shallow 1000次", () => {
    for(var i=0;i<1000;i++){
      const nav = shallow(<Nav />);
      expect(nav.text()).toBe("首页");
    }
  });
  it.skip("测试 mount 1000次", () => {
    for(var i=0;i<1000;i++){
      const nav = mount(<Nav />);
      expect(nav.text()).toBe("首页");
    }
  });
  it.skip("测试 render 1000次", () => {
    for(var i=0;i<1000;i++){
      const nav = render(<Nav />);
      expect(nav.text()).toBe("首页");
    }
  });

  // it("测试三种比较方式", () => {
  //   const o = "a";
  //   const p = "a";
  //   const x = { a: { b: 3 } };
  //   const y = { a: { b: 3, c: undefined } };
  //   const m = [, 1];
  //   const n = [undefined, 1];

  //   expect(o).toBe(p);
  //   expect(o).toEqual(p);
  //   expect(o).toStrictEqual(p);
  //   expect(x).not.toBe(y);
  //   expect(x).toEqual(y);
  //   expect(x).not.toStrictEqual(y);
  //   expect(m).not.toBe(n);
  //   expect(m).toEqual(n);
  //   expect(m).toStrictEqual(n);
  // })

  it("renders correctly", () => {
    let { props } = setup();
    const tree = renderer.create(<Detail {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  })
  it('',()=>{
    console.log('')
  })
});

