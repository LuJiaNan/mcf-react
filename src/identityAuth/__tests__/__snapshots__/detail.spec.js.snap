// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`简单测试 renders correctly 1`] = `
<div
  className="ant-panel-wrapper"
>
  <div
    className="ant-spin-nested-loading"
    style={null}
  >
    <div
      className="ant-spin-container"
    >
      <div
        className="ant-panel"
      >
        <div
          className="ant-panel-body"
        >
          <div
            className="advanced-search-panel"
          >
            <form
              className="ant-form ant-form-horizontal advanced-search-form"
              onSubmit={[Function]}
            >
              <div
                className="ant-row"
                style={
                  Object {
                    "marginLeft": -10,
                    "marginRight": -10,
                  }
                }
              >
                <div
                  className="ant-col-8"
                  style={
                    Object {
                      "paddingLeft": 10,
                      "paddingRight": 10,
                    }
                  }
                >
                  <div
                    className="ant-row ant-form-item"
                  >
                    <div
                      className="ant-col-8 ant-form-item-label"
                    >
                      <label
                        className=""
                        htmlFor="dbName"
                        onClick={[Function]}
                        title="数据库名"
                      >
                        数据库名
                      </label>
                    </div>
                    <div
                      className="ant-col-16 ant-form-item-control-wrapper"
                    >
                      <div
                        className="ant-form-item-control has-success"
                      >
                        <span
                          className="ant-form-item-children"
                        >
                          <div
                            className="ant-select ant-select-enabled ant-select-allow-clear"
                            onBlur={[Function]}
                            onClick={[Function]}
                            onFocus={[Function]}
                            onMouseDown={[Function]}
                            onTouchStart={[Function]}
                          >
                            <div
                              aria-autocomplete="list"
                              aria-expanded={false}
                              aria-haspopup="true"
                              className="ant-select-selection
            ant-select-selection--single"
                              onKeyDown={[Function]}
                              role="combobox"
                              tabIndex={0}
                            >
                              <div
                                className="ant-select-selection__rendered"
                              >
                                <div
                                  className="ant-select-selection__placeholder"
                                  onClick={[Function]}
                                  onMouseDown={[Function]}
                                  style={
                                    Object {
                                      "WebkitUserSelect": "none",
                                      "display": "none",
                                      "userSelect": "none",
                                    }
                                  }
                                  unselectable="on"
                                >
                                  全部
                                </div>
                                <div
                                  className="ant-select-selection-selected-value"
                                  style={
                                    Object {
                                      "display": "block",
                                      "opacity": 1,
                                    }
                                  }
                                  title="1"
                                >
                                  1
                                </div>
                              </div>
                              <span
                                className="ant-select-selection__clear"
                                onClick={[Function]}
                                onMouseDown={[Function]}
                                style={
                                  Object {
                                    "WebkitUserSelect": "none",
                                    "userSelect": "none",
                                  }
                                }
                                unselectable="on"
                              />
                              <span
                                className="ant-select-arrow"
                                onClick={[Function]}
                                style={
                                  Object {
                                    "WebkitUserSelect": "none",
                                    "userSelect": "none",
                                  }
                                }
                                unselectable="on"
                              >
                                <b />
                              </span>
                            </div>
                          </div>
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className="ant-col-8"
                  style={
                    Object {
                      "paddingLeft": 10,
                      "paddingRight": 10,
                    }
                  }
                >
                  <div
                    className="ant-row ant-form-item"
                  >
                    <div
                      className="ant-col-8 ant-form-item-label"
                    >
                      <label
                        className=""
                        htmlFor="assetType"
                        onClick={[Function]}
                        title="资产集合类型"
                      >
                        资产集合类型
                      </label>
                    </div>
                    <div
                      className="ant-col-16 ant-form-item-control-wrapper"
                    >
                      <div
                        className="ant-form-item-control has-success"
                      >
                        <span
                          className="ant-form-item-children"
                        >
                          <div
                            className="ant-select ant-select-enabled ant-select-allow-clear"
                            onBlur={[Function]}
                            onClick={[Function]}
                            onFocus={[Function]}
                            onMouseDown={[Function]}
                            onTouchStart={[Function]}
                          >
                            <div
                              aria-autocomplete="list"
                              aria-expanded={false}
                              aria-haspopup="true"
                              className="ant-select-selection
            ant-select-selection--single"
                              onKeyDown={[Function]}
                              role="combobox"
                              tabIndex={0}
                            >
                              <div
                                className="ant-select-selection__rendered"
                              >
                                <div
                                  className="ant-select-selection__placeholder"
                                  onClick={[Function]}
                                  onMouseDown={[Function]}
                                  style={
                                    Object {
                                      "WebkitUserSelect": "none",
                                      "display": "none",
                                      "userSelect": "none",
                                    }
                                  }
                                  unselectable="on"
                                >
                                  全部
                                </div>
                                <div
                                  className="ant-select-selection-selected-value"
                                  style={
                                    Object {
                                      "display": "block",
                                      "opacity": 1,
                                    }
                                  }
                                  title="1"
                                >
                                  1
                                </div>
                              </div>
                              <span
                                className="ant-select-selection__clear"
                                onClick={[Function]}
                                onMouseDown={[Function]}
                                style={
                                  Object {
                                    "WebkitUserSelect": "none",
                                    "userSelect": "none",
                                  }
                                }
                                unselectable="on"
                              />
                              <span
                                className="ant-select-arrow"
                                onClick={[Function]}
                                style={
                                  Object {
                                    "WebkitUserSelect": "none",
                                    "userSelect": "none",
                                  }
                                }
                                unselectable="on"
                              >
                                <b />
                              </span>
                            </div>
                          </div>
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="advanced-search-toolbar"
              >
                <button
                  className="ant-btn ant-btn-primary"
                  disabled={false}
                  onClick={[Function]}
                  type="submit"
                >
                  <span>
                    搜 索
                  </span>
                </button>
                
              </div>
            </form>
          </div>
          <div
            className="button-groups"
          >
            <div
              className="ant-btn-group"
            >
              <button
                actionkey="add"
                className="ant-btn ant-btn-primary"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                type="button"
              >
                <span>
                  添 加
                </span>
              </button>
              <span
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={
                  Object {
                    "cursor": "not-allowed",
                    "display": "inline-block",
                  }
                }
              >
                <button
                  actionkey="delete"
                  className="ant-btn"
                  disabled={true}
                  onClick={[Function]}
                  style={
                    Object {
                      "pointerEvents": "none",
                    }
                  }
                  type="button"
                >
                  <span>
                    删 除
                  </span>
                </button>
              </span>
              <button
                actionkey="deleteAll"
                className="ant-btn"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                type="button"
              >
                <span>
                  清 空
                </span>
              </button>
            </div>
          </div>
          <div
            className="ant-table-wrapper"
            style={
              Object {
                "width": "100%",
              }
            }
          >
            <div
              className="ant-spin-nested-loading"
              style={null}
            >
              <div
                className="ant-spin-container"
              >
                <div
                  className="ant-table ant-table-middle ant-table-empty ant-table-scroll-position-left"
                  style={Object {}}
                >
                  <div
                    className="ant-table-content"
                  >
                    <div
                      className="ant-table-body"
                      onScroll={[Function]}
                      onWheel={[Function]}
                      style={Object {}}
                    >
                      <table
                        className=""
                        style={Object {}}
                      >
                        <colgroup>
                          <col
                            style={
                              Object {
                                "minWidth": undefined,
                                "width": undefined,
                              }
                            }
                          />
                          <col
                            style={
                              Object {
                                "minWidth": undefined,
                                "width": undefined,
                              }
                            }
                          />
                          <col
                            style={
                              Object {
                                "minWidth": undefined,
                                "width": undefined,
                              }
                            }
                          />
                          <col
                            style={
                              Object {
                                "minWidth": 100,
                                "width": 100,
                              }
                            }
                          />
                          <col
                            style={
                              Object {
                                "minWidth": 100,
                                "width": 100,
                              }
                            }
                          />
                          <col
                            style={
                              Object {
                                "minWidth": 190,
                                "width": 190,
                              }
                            }
                          />
                        </colgroup>
                        <thead
                          className="ant-table-thead"
                        >
                          <tr
                            style={
                              Object {
                                "height": null,
                              }
                            }
                          >
                            <th
                              className="ant-table-selection-column"
                            >
                              <span>
                                <div
                                  className="ant-table-selection"
                                >
                                  <label
                                    className="ant-checkbox-wrapper"
                                  >
                                    <span
                                      className="ant-checkbox ant-checkbox-disabled"
                                      style={Object {}}
                                    >
                                      <input
                                        checked={false}
                                        className="ant-checkbox-input"
                                        disabled={true}
                                        onBlur={[Function]}
                                        onChange={[Function]}
                                        onFocus={[Function]}
                                        type="checkbox"
                                      />
                                      <span
                                        className="ant-checkbox-inner"
                                      />
                                    </span>
                                  </label>
                                </div>
                              </span>
                            </th>
                            <th
                              className=""
                            >
                              <span>
                                资产集合名称
                              </span>
                            </th>
                            <th
                              className=""
                            >
                              <span>
                                资产集合类型
                              </span>
                            </th>
                            <th
                              className=""
                            >
                              <span>
                                数据库名
                              </span>
                            </th>
                            <th
                              className=""
                            >
                              <span>
                                状态
                              </span>
                            </th>
                            <th
                              className=""
                            >
                              <span>
                                操作
                              </span>
                            </th>
                          </tr>
                        </thead>
                        <tbody
                          className="ant-table-tbody"
                        />
                      </table>
                    </div>
                    <div
                      className="ant-table-placeholder"
                    >
                      No data
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
